test/contrib/test_ssltransport.py

```diff
@@ -0,0 +1,393 @@
+class SingleTLSLayerTestCase(SocketDummyServerTestCase):
+    """
+    Uses the SocketDummyServer to validate a single TLS layer can be
+    established through the SSLTransport.
+    """
+
+    @classmethod
+    def setup_class(cls):
+        cls.server_context, cls.client_context = get_server_client_ssl_contexts()
+
+    def start_dummy_server(self):
+        def socket_handler(listener):
+            sock = listener.accept()[0]
+            with self.server_context.wrap_socket(sock, server_side=True) as ssock:
+                request = consume_socket(ssock)
+                assert request is not None
+                assert "www.testing.com" in request.decode("utf-8")
+
+                response = b"HTTP/1.1 200 OK\r\n" b"Content-Length: 0\r\n" b"\r\n"
+                ssock.send(response)
+
+        self._start_server(socket_handler)
+
+    def test_start_closed_socket(self):
+        """ Errors generated from an unconnected socket should bubble up."""
+        sock = socket.socket(socket.AF_INET)
+        context = ssl.create_default_context()
+        sock.close()
+        with pytest.raises(OSError):
+            SSLTransport(sock, context)
+
+    def test_close_after_handshake(self):
+        """ Socket errors should be bubbled up """
+        self.start_dummy_server()
+
+        sock = socket.create_connection((self.host, self.port))
+        with SSLTransport(
+            sock, self.client_context, server_hostname="localhost"
+        ) as ssock:
+            ssock.close()
+            with pytest.raises(OSError):
+                ssock.send(b"blaaargh")
+
+    def test_wrap_existing_socket(self):
+        """ Validates a single TLS layer can be established.  """
+        self.start_dummy_server()
+
+        sock = socket.create_connection((self.host, self.port))
+        with SSLTransport(
+            sock, self.client_context, server_hostname="localhost"
+        ) as ssock:
+            assert ssock.version() is not None
+            ssock.send(
+                b"GET http://www.testing.com/ HTTP/1.1\r\n"
+                b"Host: www.testing.com\r\n"
+                b"User-Agent: awesome-test\r\n"
+                b"\r\n"
+            )
+            response = consume_socket(ssock)
+            assert response is not None
+
+    def test_ssl_object_attributes(self):
+        """ Ensures common ssl attributes are exposed """
+        self.start_dummy_server()
+
+        sock = socket.create_connection((self.host, self.port))
+        with SSLTransport(
+            sock, self.client_context, server_hostname="localhost"
+        ) as ssock:
+            assert ssock.cipher() is not None
+            assert ssock.selected_alpn_protocol() is None
+            assert ssock.selected_npn_protocol() is None
+            assert ssock.shared_ciphers() is not None
+            assert ssock.compression() is None
+            assert ssock.getpeercert() is not None
```

Comment:
Will have to add test cases for `binary_form=True` and `False`, would be nice to change the `is not None` cases to type checks.

Reply:
I added ``binary_form`` as well as extended the checks done.